tests/geometry/test_parameterisations.py

```diff
@@ -189,23 +190,66 @@ def test_ordering(self):
 
 class TestComplexPictureFrame:
     @pytest.mark.parametrize(
-        "upper, lower, result",
+        "upper, lower, inner, result",
         [
-            ["CURVED", "CURVED", 57.6308],
-            ["CURVED", "FLAT", 56.829],
-            ["FLAT", "CURVED", 56.829],
+            ["CURVED", "CURVED", "TAPERED_INNER", 56.331],
+            ["CURVED", "FLAT", "TAPERED_INNER", 54.714],
+            ["FLAT", "CURVED", "TAPERED_INNER", 54.714],
+            [
+                PFrameSection.CURVED,
+                PFrameSection.CURVED,
+                PFrameSection.TAPERED_INNER,
+                56.331,
+            ],
+            [
+                PFrameSection.CURVED,
+                PFrameSection.FLAT,
+                PFrameSection.TAPERED_INNER,
+                54.714,
+            ],
+            [
+                PFrameSection.FLAT,
+                PFrameSection.CURVED,
+                PFrameSection.TAPERED_INNER,
+                54.714,
+            ],
+            ["CURVED", "CURVED", None, 57.6308],
+            ["CURVED", "FLAT", None, 56.014],
+            ["FLAT", "CURVED", None, 56.014],
+            [PFrameSection.CURVED, PFrameSection.CURVED, None, 57.6308],
+            [PFrameSection.CURVED, PFrameSection.FLAT, None, 56.014],
+            [PFrameSection.FLAT, PFrameSection.CURVED, None, 56.014],
         ],
     )
-    def test_length(self, upper, lower, result):
-        p = PictureFrame(upper=upper, lower=lower)
+    def test_length(self, upper, lower, inner, result):
+        p = PictureFrame(upper=upper, lower=lower, inner=inner)
         wire = p.create_shape()
         assert np.isclose(wire.length, result, rtol=1e-4, atol=1e-5)
 
     @pytest.mark.parametrize(
-        "upper, lower", [["CURVED", "CURVED"], ["CURVED", "FLAT"], ["FLAT", "CURVED"]]
+        "upper, lower, inner",
+        [
+            # should these pass?
+            # ["FLAT", "FLAT", "TAPERED_INNER"],
+            # ["CURVED", "CURVED", "TAPERED_INNER"],
+            # ["CURVED", "FLAT", "TAPERED_INNER"],
+            # ["FLAT", "CURVED", "TAPERED_INNER"],
+            # [PFrameSection.FLAT, PFrameSection.FLAT, PFrameSection.TAPERED_INNER],
+            # [PFrameSection.CURVED, PFrameSection.CURVED, PFrameSection.TAPERED_INNER],
+            # [PFrameSection.CURVED, PFrameSection.FLAT, PFrameSection.TAPERED_INNER],
+            # [PFrameSection.FLAT, PFrameSection.CURVED, PFrameSection.TAPERED_INNER],
```

Comment:
Should the tapered inner be tangent with the rest of the shape (it wasnt already because `["FLAT", "FLAT", "TAPERED_INNER"]` should pass and it doesnt locally

Reply:
I freely confess to not being familiar with these variants on the `PictureFrame` parameterisations. If I look at #1287, I get a better idea but I still can't be certain.

It feels to me that if the inner leg is tapered and the top and bottom legs are flat, the joints between the two cannot be tangent. They will either be a perpendicular joint, or if the tapering doesn't end before the transition into the top/bottom, it will be an obtuse angle joint.

If, on the other hand, there is some transition radius intended between the tapered leg and the flat sections, then there should be circle arcs joining up the two inner leg and the top/bottom. In this case, I would expect a tangential join. From the implementation, I vaguely recall Shail having problems getting the circle arcs to line up, hence (possibly) why there is some discretisation of these transitions along the way. It's possible this discretisation is the reason why this test fails.

If you can confirm this I can think of a few possible fixes with primitive geometry.